File: elixir/22-GenerateParenthesis/main.exs
defmodule Solution do
  @spec generate_parenthesis(n :: integer) :: [String.t]
  def generate_parenthesis(n) do
    recurse("", 0, 0, n, [])
  end

  defp recurse(current, open, close, n, acc) when open == n and close == n do
    [current | acc]
  end

  defp recurse(current, open, close, n, acc) do
    acc = if open < n do
            recurse(current <> "(", open + 1, close, n, acc)
          else
            acc
          end

    acc = if close < open do
            recurse(current <> ")", open, close + 1, n, acc)
          else
            acc
          end
    acc
  end
end

IO.inspect(Solution.generate_parenthesis(1))
IO.inspect(Solution.generate_parenthesis(3))

File: cpp/22-GenerateParenthesis/main.cpp
#include <vector>
#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) 
    {
        vector<string> store;
        recurse("", store, 0, 0, n);
        
        return store;
    }

    void recurse(string current, vector<string> &store, int open, int close, int n) const
    {
        if(open == n && close == n)
        {
            store.push_back(current);
            return;
        }

        if(open < n)
        {
            recurse(current + '(', store, open + 1, close, n);
        }

        if(close < open)
        {
            recurse(current + ')', store, open, close + 1, n);
        }
    }
};

template<typename T>
void print(const std::vector<T> &items)
{
    for(const auto &item : items)
    {
        std::cout << item << ' ';
    }

    std::cout << '\n';
}

int main()
{
    Solution solution;
    print(solution.generateParenthesis(3));
    print(solution.generateParenthesis(1));
}